API_DESIGN.md

```diff
@@ -0,0 +1,344 @@
+  // the list is sorted descending. The default is descending.
+  bool asc = 3;
+}
+```
+On the corresponding responses the `ListDetails` can be used to return the total count of the resources
+and allow the user to handle their offset and limit accordingly.
+
+
+## Error Handling
+
+The API returns machine-readable errors in the response body. This includes a status code, an error code and possibly 
+some details about the error. See the following sections for more information about the status codes, error codes and error messages.
+
+### Status Codes
+
+The API uses status codes to indicate the status of a request. Depending on the protocol used to call the API,
+the status code is returned as an HTTP status code or as a gRPC / connectRPC status code.
+Check the possible status codes https://zitadel.com/docs/apis/statuscodes
+
+### Error Codes
+
+Additionally to the status code, the API returns unique error codes for each type of error.
+The error codes are used to identify a specific error and can be used to handle the error programmatically.
+
+> [!NOTE]
+> Currently, ZITADEL might already return some error codes. However, they do not follow a specific pattern yet
+> and are not documented. We will update the error codes and document them in the future.
+
+### Error Message and Details
+
+The API returns additional details about the error in the response body.
+This includes a human-readable error message and additional information that can help the client to understand the error
+as well as machine-readable details that can be used to handle the error programmatically.
+Error details use the Google RPC error details format: https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto
+
+### Example
+
+HTTP/1.1 example:
+```
+HTTP/1.1 400 Bad Request
+Content-Type: application/json
+
+{
+  "code": "user_missing_information",
+  "message": "missing required information for the creation of the user",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.BadRequest",
+      "fieldViolations": [
+        {
+          "field": "given_name",
+          "description": "given name is required"
+        },
+        {
+          "field": "family_name",
+          "description": "family name is required"
+        }
+      ]
+    }
+  ]
+}
+```
+
+gRPC / connectRPC example:
+```
+HTTP/2.0 200 OK
+Content-Type: application/grpc
+Grpc-Message: missing required information for the creation of the user
+Grpc-Status: 3
+
+{
+  "code": "user_missing_information",
+  "message": "missing required information for the creation of the user",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.BadRequest",
+      "fieldViolations": [
+        {
+          "field": "given_name",
+          "description": "given name is required"
+        },
+        {
+          "field": "family_name",
+          "description": "family name is required"
+        }
+      ]
+    }
+  ]
+}
```

Comment:
Given the examples my assumption is that we return different errors for missing infomration and invalid information. E.g. username too long would be a different error code than no user name. Is that correct?

Reply:
Yes. Generally, there should be a seperate code for different reasons of the error, resp. if it (potentially) needs to be treated differently. I guess though it will be hard to create a strict rule when exactly there needs be a seperate error or not.
In this specific case i would sperate them, as we (also only currently) have the requirement that these specific fields need to be set. Whereas the length of different fields are more general and could be treated in a general error code, maybe even not user specific?

Reply:
Im unsure if I care about the differentiation as a user as long as the details clearly state the violation. I would prefer to have a single error code if the input is invalid and then go through the details instead of handling multiple codes which require the same action on my side (recheck input, change, resend).  I also don't care which LOC in Zitadel caused the error. 

So for me it would make more sense to have unique codes for the action I need to take and find the information in the details. If the validation of the payload fails I want a single code for that, if Zitadels business logic fails, because the username is already taken I want a different error code because I need to handle the error differently.

Reply:
totally agree on the LOC part

as discussed, i updated the example